### PR TITLE
KAFKA-13971:Atomicity violations caused by improper usage of ConcurrentHashMap

### PR DESCRIPTION
**Pull request by Kvicii**
ensure thread safe

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

### CODE REVIEW

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfigTransformer.java [hunk 1]
```diff
@@ -81,7 +81,7 @@ private void scheduleReload(String connectorName, String path, long ttl) {
         Map<String, HerderRequest> connectorRequests = requests.get(connectorName);
         if (connectorRequests == null) {
             connectorRequests = new ConcurrentHashMap<>();
-            requests.put(connectorName, connectorRequests);
+            requests.putIfAbsent(connectorName, connectorRequests);
```
Comment (divijvaidya) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfigTransformer.java]: This is still incorrect. Consider the following scenario:

Thread T1 arrives and reaches line 83. Context switch occurs to thread T2.
T2 arrives and also reaches line 83. Context switch occurs to thread T1.
T1 executes line 99 and adds an entry. Context switch occurs to thread T2.
In a world where we want serialized execution, T2 should have gone to the else block, cancelled the previous request and added the new request but now, T2 will simply overwrite the request at line 99 without having cancelled the previous one.

Comment (divijvaidya) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfigTransformer.java:84]: You should probably update this code as follows:
```
private void scheduleReload(String connectorName, String path, long ttl) {
        final Map<String, HerderRequest> connectorRequests = requests.computeIfAbsent(connectorName, (c) -> new ConcurrentHashMap<>());
        connectorRequests.computeIfPresent(path, (p, previousRequest) -> {
            // Delete previous request for ttl which is now stale
            previousRequest.cancel();

            log.info("Scheduling a restart of connector {} in {} ms", connectorName, ttl);
            Callback<Void> cb = (error, result) -> {
                if (error != null) {
                    log.error("Unexpected error during connector restart: ", error);
                }
            };
            return worker.herder().restartConnector(ttl, connectorName, cb);
        });
    }
```

Comment (divijvaidya) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfigTransformer.java]: You might want to use code similar to the following:
```
    private void scheduleReload(String connectorName, String path, long ttl) {
        final Map<String, HerderRequest> connectorRequests = requests.computeIfAbsent(connectorName, (c) -> new ConcurrentHashMap<>());
        connectorRequests.compute(path, (p, previousRequest) -> {
            if (previousRequest != null) {
                // Delete previous request for ttl which is now stale
                previousRequest.cancel();
            }

            log.info("Scheduling a restart of connector {} in {} ms", connectorName, ttl);
            Callback<Void> cb = (error, result) -> {
                if (error != null) {
                    log.error("Unexpected error during connector restart: ", error);
                }
            };
            return worker.herder().restartConnector(ttl, connectorName, cb);
        });
    }
```

**Review by divijvaidya — APPROVED** [review]
Thanks for fixing the code as per the suggestion. LGTM!

**Comment by Kvicii** [comment]
@hachikuji can you help me review this PR?

**Comment by ijuma** [comment]
We should check if this code is used in a concurrent context. Will leave it to @C0urante.

**Review by C0urante — APPROVED** [review]
It's unlikely but definitely possible for `scheduleReload` to be invoked concurrently, since tasks are allowed to access their own configs through their `SourceTaskContext`/`SinkTaskContext`.

I wish we could have a test for this but considering how tricky it is to reliably reproduce race conditions and how simple the changes are, I'm comfortable making an exception and merging this change without a unit test.

LGTM, thanks @Kvicii!

**Pull request merged**